### PR TITLE
feat(returns): opt-in per-group breakdown for report returns (--by-group)

### PR DESCRIPTION
**Pull request by robcohen**
## What

Adds an opt-in per-group breakdown to `rledger report returns` (#1820). Tag `open` directives with `returns-group: "Name"` and pass `--by-group`:

```beancount
2020-01-01 open Assets:Broker:AAPL
  returns-group: "Tech"
2020-01-01 open Income:Dividends
  returns-group: "Tech"     ; makes Tech dividend-inclusive
```

```
Group                                 MWR      TWR    Invested     Current
------------------------------------------------------------------------
Bonds                              10.00%   10.00%         500         550
Tech                               32.36%   32.60%        1000        1300
------------------------------------------------------------------------
TOTAL                              24.85%   24.98%        1500        1850
```

Each group is an **independent sub-portfolio**: its return is computed over just that group's accounts, exactly as if you'd narrowed `--investments`/`--income` to them. This is the model beangrow and hledger `roi` use. It follows the prior-art direction (declared groups that bundle their income accounts, so the return is dividend-inclusive — declared in the ledger, not beangrow's external config file).

## Design: groups are reported alongside the total, not as a partition of it

An earlier revision promised the group rows summed to the TOTAL (an `(ungrouped)` residual). A deep review proved that guarantee **unsound**: when two groups draw on a shared in-scope account (a pooled settlement-cash account under `--investments` — the common brokerage shape), the boundary flow into that shared account can't be attributed to one group. This is the same un-attributability that made us drop `--by-account`. So:

- **No `(ungrouped)` residual.** Only tagged accounts appear; an untagged in-scope holding is omitted from the breakdown but still counted in the TOTAL. The TOTAL is the whole-portfolio figure, reported for reference — not claimed as the sum of the rows.
- **Warnings (stderr), not silent misreporting**, for the cases a reader would misread:
  - a group that is **not self-contained** (shares an in-scope account with the rest of the portfolio — its return then counts an internal transfer as a flow);
  - two groups whose accounts **overlap by prefix** (a tagged parent or duplicate `open` values the same holding in both);
  - a `returns-group:` tag on an **out-of-scope** account (Equity/Liability, or outside `--investments`/`--income`) — dropped, never valued as a phantom holding;
  - a **non-string** tag value.
- **Why not auto per-account/per-commodity:** a dividend booked to a shared cash/income account can't be attributed to one holding automatically — which is exactly why every reference tool uses declared groups.

## Tests

Independent-groups (no residual; TOTAL ≠ sum of rows), self-contained warning, out-of-scope + non-string tag warnings, income-only (empty-investment) group, grouped **text** path, and a positively-pinned default-shape guard. Plus the engine's existing suite and the two drift-guard unit tests.

## Deferred (per #1820)

True per-commodity attribution within a single account, `commodity`-directive tagging, and the N+1-extraction perf pass (#1832).

Part of #1820.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by robcohen** [comment]
## Rework — deep-review findings addressed

Reworked the grouping design per the deep review (2 decisions confirmed with @robcohen: drop `--by-account`; make grouping an opt-in flag with an ungrouped residual). Mapping to the 6 findings:

| # | Finding | Resolution |
|---|---------|-----------|
| 1 | `--by-account` with empty income wrongly booked a dividend-into-holding as a contribution (invested 1020) | **`--by-account` dropped** — per-account auto-attribution is unsound |
| 2 | Metadata groups classified independently of `--investments`, so they didn't reconcile with TOTAL | Groups now classified against the **whole reporting scope**; untagged in-scope holdings collect into an `(ungrouped)` residual → rows sum to TOTAL |
| 3 | Equity/Liability tagged → routed into the investment scope, valued as holdings | An out-of-scope tag (External / Equity / Liability) is **ignored with a warning** |
| 4 | Merely having `returns-group:` metadata silently switched the output schema | Grouping is **opt-in behind `--by-group`**; default output shape unchanged |
| 5 | A non-string `returns-group:` value was silently dropped | Now **warns** |
| 6 | N+1 extractions per group (perf) | **Deferred to #1832** (the twr O(n) pass), noted there |

New/changed tests: `--by-group` dividend-inclusive grouping, `(ungrouped)` residual + TOTAL reconciliation (invested 1000+500=1500, current 1300+550=1850), and default-is-single-summary. Gauntlet green (clippy/doc/fmt/tests).

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 1]
```diff
@@ -181,42 +327,162 @@ pub(super) fn report_returns<W: Write>(
...
+                    w,
+                    "{:32}{:>8}%{:>8}%{:>12}{:>12}",
+                    truncate(&r.label, 32),
+                    fmt_rate(r.mwr),
+                    fmt_rate(r.twr),
+                    money(r.invested),
+                    money(r.current_value),
+                )?;
+                Ok(())
+            };
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs:636]: Fixed in abba4689. Added `fmt_rate_pct` for the grouped text table (the value carries its own `%`; `None` → `n/a`), matching `render_single`'s single-scope output. CSV/JSON keep the bare numeric. Pinned with a unit test (`undefined_rate_renders_without_a_percent_sign`).

**Comment by robcohen** [comment]
## Second deep review → reworked to independent sub-portfolios

A high-effort review found the "(ungrouped) residual + rows reconcile with TOTAL" model was **unsound**, and I reproduced the two architectural cases against the real binary before acting:

**Finding 1 — shared in-scope cash (common brokerage shape):** with `Assets:Broker:Cash` under `--investments`, per-sub-scope extraction counts the pooled cash's funding against every group → invested rows summed to 3000 vs a 1500 TOTAL. **Fundamental**: a boundary flow into a shared account can't be attributed to one group (same un-attributability that killed `--by-account`).

**Finding 4 — prefix overlap:** an opened parent `Assets:Broker` prefix-matched its tagged children, so `(ungrouped)` re-counted the whole portfolio.

Decision (confirmed): **drop the reconciliation promise; present each group as an independent sub-portfolio** (the beangrow / hledger `roi` model). Mapping of all 9 findings:

| # | Finding | Resolution |
|---|---------|-----------|
| 1 | Shared in-scope cash → phantom flows, no reconciliation | Groups are now independent sub-portfolios; a non-self-contained group **warns** naming the shared account |
| 2 | Membership from `open`s but TOTAL from postings → posted-but-unopened vanishes | Dissolved: only tagged accounts appear; no residual claiming to cover the scope |
| 3 | Duplicate `open`s, different tags → double count | **Prefix-overlap warning** (covers exact-duplicate + ancestor cases) |
| 4 | Parent-prefix `open` swallows children | Untagged-parent case dissolved (no residual); tagged-parent overlap **warns** |
| 5 | `(ungrouped)` label collision | Dissolved: residual removed |
| 6 | Text `--by-group` path untested | Added `by_group_text_output_renders_rows_and_total` |
| 7 | No warning-path / empty-investment-group tests | Added self-contained, out-of-scope, non-string, and income-only-group tests |
| 8 | N+1 extraction | Deferred to #1832 (noted) |
| 9 | Weak `!contains("groups")` default test | Now positively pins the single-summary schema |

Pushed in `2211d2e3`. Gauntlet green (clippy/doc/fmt; 14 integration + 3 unit tests). The TOTAL row is now documented as the whole-portfolio figure, reported alongside the groups rather than as their sum.

**Comment by robcohen** [comment]
## Third deep review — polish (reconciliation issues did not recur)

A high-effort review of the reworked design found 10 findings (1 refuted). None re-surfaced the reconciliation problems — the independent-sub-portfolio model holds. Resolutions:

| # | Finding | Resolution |
|---|---------|-----------|
| F2 | Income-only group reported a **fabricated 0.00% TWR** (`twr()` returns `Some(0.0)` over an empty position) | TWR is `None`/`n/a` when a scope never held capital (`0fa6d3ca`); test asserts it |
| F7 | Text table rules were 72 cols but rows 74 → last column overhung | Rule width derived from the column layout |
| F5 | A newline in a `returns-group:` value could split/inject a text row | Control chars sanitized in the text label (JSON/CSV already escape) |
| F4 | A group named `TOTAL` collides with the total row in text/CSV | Warns (JSON keeps them structurally separate) |
| F9 | Prefix-overlap warning untested | Added `by_group_warns_on_cross_group_prefix_overlap` |
| F10 | Self-contained test discarded stdout | Now asserts the warned group still renders |
| F6 | Grouped text dropped the Distributions column | Restored for parity with single-scope (`5ba3f5f7`) |
| F8 | `TOTAL` row read like a subtotal | Footnote: "TOTAL is the whole portfolio, not the sum of the groups" |
| F1/F3 | Overlap / not-self-contained warnings are stderr-only | **Kept stderr-only by design**: for an independent sub-portfolio the figure is a legitimate standalone view with a caveat, so a per-format `caveats` schema field would be disproportionate |
| — | (refuted) CSV per-row vs JSON top-level `as_of`/`currency` | Acceptable per-format shaping, not a defect |

Gauntlet green (clippy/doc/fmt; 15 integration + 3 unit tests). This is the third review pass; the reconciliation model has been stable across the last two.

**Comment by robcohen** [comment]
## Fourth deep review — two real correctness bugs found + fixed

Scoped to hunt only for NEW defects (the eight prior resolutions were excluded). Both correctness findings reproduced live before fixing; fixed in `556d243d`:

| # | Sev | Finding | Fix |
|---|-----|---------|-----|
| F1 | 🔴 Correctness | `--by-group` emitted **two incompatible schemas** — with no in-scope `returns-group:` tags it fell back to the single-summary object (no `groups`/`total`), breaking JSON consumers | Always emits the grouped schema (`"groups": []` + `total`); the notice stays on stderr |
| F2 | 🔴 Correctness | A raw control byte (e.g. `U+001B`) in a `returns-group:` value produced **invalid JSON** — report's `json_escape` was aliased to `escape_string`, the beancount-*source* escaper (only `" \ \n \t \r`) | Added a proper `escape_json` to core (RFC 8259: `\b \f \uXXXX`); rewired report's `json_escape`. Fixes **all** report JSON surfaces, not just groups |
| F6 | 🟡 Hardening | Text-label sanitizer missed `U+2028`/`U+2029` (line/paragraph separators) | Neutralized alongside control chars in `truncate` |

**Deliberately deferred (tool-wide, not returns-specific — worth a separate decision):**
- **F3** CSV formula-injection (`=`/`+`/`-`/`@` prefix): `escape_csv` is the shared canonical (BQL CSV too); guarding only this column would be inconsistent, and a global policy change shouldn't ride in on this PR.
- **F4/F5** fixed-width table alignment — a monetary value wider than its column, and Unicode display-width (CJK/combining) counted as scalars. General table-rendering limitations shared across the tool; truncating a money value would be worse than misaligning.
- **F7** the self-containment scan is a second O(groups×txns) pass → folded into the deferred perf work (#1832).

Tests added: `escape_json` unit test, `--by-group`-with-no-tags grouped-schema test, control-char-label valid-JSON test. Gauntlet green (429 core + 17 returns tests, clippy/doc/fmt).

Four review passes now; the correctness core has been stable since pass 2, and this pass's findings were output-format/schema hygiene rather than model defects.

**Comment by robcohen** [comment]
## Fifth deep review — converged (2 minor follow-ups)

As expected, yield has flattened. Both findings were consequences of the fourth-pass fixes, both fixed in `20cb1fc0`:

- **Text layout:** the no-tags `--by-group` text output printed two horizontal rules back-to-back (no group rows between the header rule and the pre-TOTAL rule). Now skips the separator when there are no group rows. Test added.
- **Stale comment:** the comment above the `json_escape` alias still described `escape_string`/behavior-identity after the switch to `escape_json`. Corrected, with a do-not-revert note.

No correctness or model defects this pass. Gauntlet green (17 returns + core tests, clippy/doc/fmt). Five passes in, the feature has converged.

**Comment by robcohen** [comment]
## Copilot review threads — resolutions

Addressing the outstanding inline threads (fixed in `1bd649f9`):

**Horizon bounding (the substantive catch — thank you).** `build_groups` iterated every `open` and `first_shared_inscope_account` scanned every transaction regardless of `--end`, while flow extraction/valuation are horizon-bounded. A historical report (`--end` in the past) therefore showed spurious all-zero rows for groups whose accounts open later, and could raise false self-containment warnings from post-horizon transfers. **Fixed:** both now take `end_date` and skip directives dated after it. Reproduced (a `Crypto` group opened in 2022 no longer appears in a 2020 report) and pinned with `by_group_respects_the_end_horizon`.

**stderr warning injection (group name).** Warnings interpolated the raw, user-controlled group name. **Fixed:** factored the control-char / U+2028-U+2029 neutralization out of `truncate()` into `sanitize_display()` and applied it to the stderr warnings.

**CSV control characters in the label.** **Fixed:** the CSV label now passes through `sanitize_display()` before `csv_escape` (JSON keeps `escape_json`, which is lossless). Note: CSV *formula*-injection (`=`/`+`/`-`/`@`) remains a tool-wide policy for the shared `escape_csv` and is intentionally out of scope for this PR.

**`n/a%` in grouped text (outdated).** Already fixed earlier in this PR via `fmt_rate_pct` (undefined rate renders `n/a`, not `n/a%`).

**Centralizing `returns-group` metadata parsing in core.** There is no existing canonical metadata-string getter to reuse (checked); the parse here is a single `MetaValue::String` match with two typed warnings. Extracting a core abstraction for one call site isn't warranted — will revisit if a second consumer needs the same tag.

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 2]
```diff
@@ -181,42 +494,184 @@ pub(super) fn report_returns<W: Write>(
...
+                    currency,
+                    r.flow_count,
+                    csv_escape(&money(r.invested)),
+                    csv_escape(&money(r.distributions)),
+                    csv_escape(&money(r.current_value)),
+                    fmt_rate(r.mwr),
+                    fmt_rate(r.twr),
+                )?;
+            }
+        }
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs:587]: By design. JSON keeps the label losslessly via escape_json (RFC 8259 \uXXXX). CSV and text have no standard lossless-and-safe escape for control bytes — CSV has no \u escape, and emitting the raw byte is exactly the terminal/viewer hazard we avoid — so those surfaces sanitize control chars to spaces. A control byte in a returns-group label is an extreme edge case (self-authored metadata); the machine-readable JSON is the source of truth and is exact.

File: crates/rustledger/tests/report_returns_test.rs [hunk 3]
```diff
@@ -349,3 +521,488 @@ fn time_weighted_return_diverges_from_money_weighted_on_timing() {
...
+        .unwrap_or_else(|| panic!("no TOTAL row: {out}"));
+    assert!(total.contains("1850"), "TOTAL row: {total}");
+    // A footnote makes clear the TOTAL is not the sum of the group rows.
+    assert!(
+        out.contains("not the sum of the groups"),
+        "expected the non-sum TOTAL footnote: {out}"
+    );
+}
+
+#[test]
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/tests/report_returns_test.rs:834]: Added by_group_csv_output_has_header_group_and_total_rows (1d8e3bf3): asserts the CSV header, a group data row, and the TOTAL row.

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 4]
```diff
@@ -34,11 +73,265 @@ impl PriceOracle for PriceDbOracle<'_> {
...
+    }
+
+    // Cross-group prefix overlap: `Scope` classifies by prefix, so a tagged
+    // account that is an ancestor of (or identical to) another group's tagged
+    // account silently values the same holding in both groups.
+    for (i, (a, group_a)) in tagged.iter().enumerate() {
+        for (b, group_b) in &tagged[i + 1..] {
+            if group_a != group_b && (is_subaccount_or_equal(a, b) || is_subaccount_or_equal(b, a))
+            {
+                let (group_a, group_b) = (sanitize_display(group_a), sanitize_display(group_b));
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs:235]: Agreed it's O(k^2) over tagged accounts (k = tagged opens, typically small — not all opens). Folding into the deferred returns perf pass (#1832) alongside the N+1 per-group extraction and the self-containment scan; not worth a trie for the common handful-of-groups case, and correctness is unaffected.

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 5]
```diff
@@ -34,11 +73,265 @@ impl PriceOracle for PriceDbOracle<'_> {
...
+    }
+
+    rows
+}
+
+/// Replace control characters and the Unicode line/paragraph separators
+/// (U+2028/U+2029) with a space. A `returns-group:` label is arbitrary
+/// user-controlled text; on the human-facing surfaces (the text table, CSV, and
+/// stderr warnings) such bytes could inject terminal escapes or extra lines, so
+/// they are neutralized. JSON escapes them losslessly (`\uXXXX`) instead.
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs]: Fixed in b91d9728. Reworded: JSON keeps the label via escape_json — C0 bytes as \uXXXX, U+2028/U+2029 as-is (already valid in a JSON string).

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 6]
```diff
@@ -181,42 +494,184 @@ pub(super) fn report_returns<W: Write>(
...
+    }
     Ok(())
 }
 
+/// Truncate a label to fit a text column (keeps the informative tail).
+///
+/// Control characters (e.g. a newline in a quoted `returns-group:` value) are
+/// replaced with a space first, so a label can neither split the fixed-width row
+/// across lines nor inject a spoofed second line into the text report. The JSON
+/// and CSV paths escape the raw label; only the text table needs this.
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs]: Fixed in b91d9728. truncate's doc now says the label goes through sanitize_display; the raw-label-escaping claim (inaccurate for CSV, which sanitizes) is removed.

File: crates/rustledger/src/cmd/report_cmd/returns.rs [hunk 7]
```diff
@@ -34,11 +73,267 @@ impl PriceOracle for PriceDbOracle<'_> {
...
+/// - a non-string `returns-group:` value;
+/// - two groups whose accounts overlap by prefix — `Scope` matches by prefix, so
+///   a tagged ancestor (or a duplicate `open`) values another group's holding
+///   twice;
+/// - a group that is not self-contained: it shares an in-scope account with the
+///   rest of the portfolio, so its flows count an internal transfer as a
+///   contribution/withdrawal.
+///
+/// An empty result means no in-scope account carried a usable tag; the caller
+/// then reports the plain total.
```
[the remark this answers is not in the document]
Comment (robcohen) [reply crates/rustledger/src/cmd/report_cmd/returns.rs]: Fixed in 85add459: build_groups' rustdoc now says the caller emits the grouped shape (empty group list + total) on an empty result, and documents the end_date horizon bound.
